### PR TITLE
Rectangle tool

### PR DESCRIPTION
**Pull request by 1ntEgr8**

### CODE REVIEW

**Comment by Litagano-M** [comment]
When using the select tool to select shapes, the nodes for previous shapes don't disappear when you select a new shape. Is that intended?

File: js/shape.js [hunk 1]
```diff
@@ -54,7 +57,7 @@ class Shape {
             canvas.shapes[shapeId].showNodes();
```
Comment (joel99) [inline js/shape.js]: is there no onclick we can use for this shape conveniently? can you explain

File: js/main.js [hunk 2]
```diff
@@ -4,21 +4,38 @@ const canvas = new Canvas("canvas", "canvas-container");
...
+
+    rectangleButton.addEventListener("click", () => {
+        cleanUpOnSelectingNewTool();
+
+        canvas.canvas.addEventListener("click", Rectangle.onMouseClick);
+    })
+});
+
+function cleanUpOnSelectingNewTool() {
+    canvas.clickToggle = true;
```
Comment (joel99) [inline js/main.js]: it seems fairly intensive to add and remove all these listeners all the time, if there are a lot of shapes?

I would try holding "activeShape(s)" in a canvas/MapGT package context (class variable) and have pass Map into each these shapes' constructors in order to provide each individual shape a way of registering a click handler in map scope. Then, on click on any shape, package level logic (such as what active shape/tool is) can decide the appropriate action to take - such as forwarding to the shape onClick, or activating new shapes, etc.

This is a bit vague, lmk if you want more detailing, or use your own judgment, if you think this is bad advice (have your own reasoning about what will take effort to maintain)

File: js/path.js [hunk 3]
```diff
@@ -57,7 +57,7 @@ class Path extends Shape {
     }
 
     static onMouseMove(event) {
-        let pathObject = canvas.currentShape; 
+        let pathObject = canvas.currentShape;
```
Comment (joel99) [inline js/path.js]: use an ide with linting

File: js/rectangle.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+            this.nodes[i].changeStyles({
+                "visibility": "hidden"
+            });
+            this.nodes[i].domElement.removeEventListener("mousedown", Rectangle.onMouseDownNode);
+            this.nodes[i].domElement.removeEventListener("mouseup", Rectangle.onMouseUpNode);
+        }
+    }
+
+    static onMouseClick(event) {
+        if (canvas.clickToggle) {
```
Comment (joel99) [inline js/rectangle.js]: lmk what clickToggle is

**Comment by 1ntEgr8** [comment]
> When using the select tool to select shapes, the nodes for previous shapes don't disappear when you select a new shape. Is that intended?

yea that's intended; but if it seems wrong, I can disable it

File: js/main.js (hunk 2 above)
Comment (1ntEgr8) [reply js/main.js]: I got what you're thinking; I will try to implement my understanding of it in the PR for the next tool if that's fine?

File: js/shape.js (hunk 1 above)
Comment (1ntEgr8) [reply js/shape.js]: I didn't understand :(

**Pull request merged**